src/durmod/kaldi-durmod.h

```diff
@@ -0,0 +1,266 @@
+#include "util/stl-utils.h"
+#include "fstext/deterministic-fst.h"
+
+namespace kaldi {
+  using nnet3::Nnet;
+  using nnet3::NnetExample;
+  using nnet3::NnetIo;
+  using nnet3::CachingOptimizingCompiler;
+  using nnet3::ComputationRequest;
+  using nnet3::NnetComputation;
+  using nnet3::NnetComputeOptions;
+  using nnet3::NnetComputer;
+
+struct PhoneDurationModelOptions {
+  int32 left_context, right_context;
+  int32 max_duration;
+
+  void Register(OptionsItf *po);
+
+  PhoneDurationModelOptions():
+    left_context(4),
+    right_context(2),
+    max_duration(60) {}
+};
+
+class PhoneDurationModel {
+  friend class PhoneDurationFeatureMaker;
+
+ public:
+  void Read(std::istream &is, bool binary);
+  void Write(std::ostream &os, bool binary) const;
+
+  PhoneDurationModel() {}
+  PhoneDurationModel(const PhoneDurationModelOptions &opts,
+                     const std::vector<std::vector<int32> > &roots,
+                     const std::vector<std::vector<int32> > &questions):
+    left_context_(opts.left_context),
+    right_context_(opts.right_context),
+    roots_(roots),
+    questions_(questions),
+    max_duration_(opts.max_duration) {}
+
+  inline int32 MaxDuration() const { return max_duration_; }
+  inline int32 RightContext() const { return right_context_; }
+  inline int32 LeftContext() const { return left_context_; }
+  inline int32 FullContextSize() const {
+    return left_context_ + right_context_ + 1;
+  }
+  std::string Info() const;
+
+ private:
+  int32 left_context_, right_context_;
+  std::vector<std::vector<int32> > roots_;
+  std::vector<std::vector<int32> > questions_;
+  int32 max_duration_;
+};
+
+class PhoneDurationFeatureMaker {
+ public:
+  explicit PhoneDurationFeatureMaker(const PhoneDurationModel &model);
+
+  void InitFeatureMaker(const PhoneDurationModel &model);
+
+  /// This method extracts features for a phone. The inputs are a
+  /// phone-duration context and an index into that context which indicates the
+  /// middle phone (middle in the sense of left/right context).
+  /// a phone-duration context is a sequence of <phone-ID, duration-in-frames>
+  /// pairs.
+  /// The extracted features consist of phone-IDs (which are represented in a
+  /// 1-of-n encoding, suitable for neural networks), phone durations, and
+  /// binary features determined from a set of questions
+  /// (i.e. extra_questions.int)
+  void MakeFeatureVector(
+                 const std::vector<std::pair<int32, int32> > &phone_dur_context,
+                 int phone_index,
+                 SparseVector<BaseFloat> *feat) const;
+
+
+  int32 FeatureDim() const { return feature_dim_; }
+  int32 NumBinaryFeatures() const { return num_binary_features_; }
+  int32 NumPhoneIdentities() const { return num_phone_identities_; }
+  int32 OutputDim() const { return max_duration_; }
+
+  std::string Info() const;
+
+ private:
+  unordered_map<int32, std::vector<int32> > binary_feats_;
+  unordered_map<int32, int32> phone_id_;
```

Comment:
OK.
BTW, if it looks (e.g. on other datasets) like this
subtracting-average-log-likelihood thing is working, then there is a
cleaner way to do it, using the neural net.   This is just an idea- let me
know what you think.
The idea is we create another 'branch' of the neural net topology that only
sees the input phones but not the input durations (easy to do with
dim-range-node, if the durations appear at the end of the feature vector).
  And we create a differently-named output, called, for example,
average-logprob.  And during training we train that with an objective
function that's -0.5 * (average-logrob - log(prob at output, of correct
duration)).   So it learns a mapping from the phone-in-context to the
average logprob.  This means we don't have to mess with backoff, etc.
Of course, if the bigram logprob subtraction is no better than the
monphone, or only slightly better, this is probably overkill.

Dan


On Thu, Feb 18, 2016 at 5:36 PM, Hossein Hadian <notifications@github.com>
wrote:

> In src/durmod/kaldi-durmod.h
> <https://github.com/kaldi-asr/kaldi/pull/414#discussion_r53394431>:
>
> > +  void MakeFeatureVector(
> > +                 const std::vector<std::pair<int32, int32> > &phone_dur_context,
> > +                 int phone_index,
> > +                 SparseVector<BaseFloat> *feat) const;
> > +
> > +
> > +  int32 FeatureDim() const { return feature_dim_; }
> > +  int32 NumBinaryFeatures() const { return num_binary_features_; }
> > +  int32 NumPhoneIdentities() const { return num_phone_identities_; }
> > +  int32 OutputDim() const { return max_duration_; }
> > +
> > +  std::string Info() const;
> > +
> > + private:
> > +  unordered_map<int32, std::vector<int32> > binary_feats_;
> > +  unordered_map<int32, int32> phone_id_;
>
> In fact these mappings were present in the nnet2 partial implementation
> too (by Yenda). I will fix the name and add documentation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/414/files#r53394431>.
>